File: client/browser/js/agency.js
// ═══════════════════════════════════════════════════════
// līlā — Client-Side Agency Engine
//
// Between server ticks, each mobile entity decides what to do based on:
//   - Server intent (state + drives + eligibility flags)
//   - Local perception (nearest food, water, threats from world model)
//   - Motion latent (modulates speed, hesitation, path curvature)
//
// This is the "body" in "server is nervous system, client is body."
// ═══════════════════════════════════════════════════════

import { GRID_SIZE } from './constants.js';

/**
 * Run one frame of local agency for all mobile entities.
 * Called every render frame (~60 Hz).
 */
export function stepAgency(world, dt) {
  const events = []; // client-reported events to send upstream

  for (const ent of world.entities.values()) {
    if (!ent.isMobileConsumer || !ent.isAlive) continue;

    // Update speed from species definition (once)
    if (!ent._speedSet) {
      const def = world.getSpeciesDef(ent.species);
      ent.speed = def?.movement_speed ?? 2.0;
      ent._speedSet = true;
    }

    // Evaluate behavior based on state + drives + eligibility
    const action = evaluateBehavior(ent, world);

    // Execute the action (move toward target, interact, etc.)
    executeAction(ent, action, world, dt, events);
  }

  return events;
}

/**
 * Evaluate what an entity should do based on its intent and local perception.
 * Returns { type, target? } describing the desired action.
 */
function evaluateBehavior(ent, world) {
  const state = ent.state;
  const drive = ent.drive || {};
  const speciesDef = world.getSpeciesDef(ent.species);

  // ── Fleeing (highest priority — threat detected locally) ──
  if (state === 'FLEEING') {
    return evaluateFleeing(ent, world, speciesDef);
  }

  // ── Drinking ──
  if (state === 'DRINKING' || (ent.canDrink && drive.hydration < 0.3)) {
    return evaluateDrinking(ent, world);
  }

  // ── Reproduction seeking ──
  if (ent.reproEligible && drive.reproductive_drive > 0.5) {
    const mateAction = evaluateMateSeeking(ent, world);
    if (mateAction) return mateAction;
  }

  // ── Foraging / Herbivory ──
  if (state === 'FORAGING' && ent.canConsume) {
    return evaluateForaging(ent, world, speciesDef);
  }

  // ── Hunting / Predation ──
  if ((state === 'HUNTING' || state === 'FORAGING') && ent.canPredatate) {
    return evaluateHunting(ent, world, speciesDef);
  }

  // ── Pollination ──
  if (ent.canPollinate && speciesDef?.is_pollinator) {
    return evaluatePollination(ent, world, speciesDef);
  }

  // ── Resting / Idle — wander with latent-modulated style ──
  return evaluateWandering(ent, world);
}

// ─── Behavior Evaluators ──────────────────────────────

function evaluateFleeing(ent, world, speciesDef) {
  const fleeTargets = speciesDef?.flee_targets || [];
  if (fleeTargets.length === 0) return { type: 'wander' };

  // Find nearest threat
  let nearestThreat = null;
  let bestDist = Infinity;
  for (const other of world.entities.values()) {
    if (!other.isAlive) continue;
    const def = world.getSpeciesDef(other.species);
    if (!def || !fleeTargets.includes(other.species)) continue;
    const d2 = ent.distSqTo(other);
    if (d2 < bestDist) {
      bestDist = d2;
      nearestThreat = other;
    }
  }

  if (nearestThreat && bestDist < 400) { // ~20 world units sensory range²
    // Flee away from threat
    const dx = ent.x - nearestThreat.x;
    const dz = ent.z - nearestThreat.z;
    const dist = Math.sqrt(dx * dx + dz * dz) || 1;
    return {
      type: 'flee',
      targetX: clamp(ent.x + (dx / dist) * 8, GRID_SIZE),
      targetZ: clamp(ent.z + (dz / dist) * 8, GRID_SIZE),
    };
  }

  // No threat nearby — fall through to wander
  return { type: 'wander' };
}

function evaluateDrinking(ent, world) {
  const water = world.findNearestWater(ent.x, ent.z);
  if (water) {
    // Approach water source edge
    const wx = water.position[0], wz = water.position[2];
    const r = water.radius || 1;
    const dx = wx - ent.x, dz = wz - ent.z;
    const dist = Math.sqrt(dx * dx + dz * dz) || 1;
    // Approach to just outside the water edge
    const approachR = Math.max(r - 0.5, 0.3);
    return {
      type: 'drink',
      targetX: clamp(wx - (dx / dist) * approachR, GRID_SIZE),
      targetZ: clamp(wz - (dz / dist) * approachR, GRID_SIZE),
    };
  }
  return { type: 'wander' }; // no water found
}

function evaluateMateSeeking(ent, world) {
  const mate = world.findNearestMate(ent);
  if (mate && ent.distanceTo(mate) < 15) {
    return {
      type: 'seek_mate',
      targetX: mate.x,
      targetZ: mate.z,
      targetId: mate.id,
    };
  }
  return null; // no mate nearby, fall through to other behaviors
}

function evaluateForaging(ent, world, speciesDef) {
  const dietOrder = speciesDef?.diet_order || [];

  // Try each food preference in order
  for (const [foodSpecies] of dietOrder) {
    const food = world.findNearestSpecies(ent.x, ent.z, [foodSpecies], ent.id);
    if (food && ent.distanceTo(food) < 15) {
      return {
        type: 'forage',
        targetX: food.x,
        targetZ: food.z,
        targetId: food.id,
      };
    }
  }

  // No preferred food — wander to search
  return { type: 'wander' };
}

function evaluateHunting(ent, world, speciesDef) {
  const dietOrder = speciesDef?.diet_order || [];
  const preySpecies = dietOrder.map(([s]) => s);

  if (preySpecies.length > 0) {
    const prey = world.findNearestSpecies(ent.x, ent.z, preySpecies, ent.id);
    if (prey && ent.distanceTo(prey) < 15) {
      return {
        type: 'hunt',
        targetX: prey.x,
        targetZ: prey.z,
        targetId: prey.id,
      };
    }
  }

  // No prey — wander (or fall back to foraging if omnivore)
  return { type: 'wander' };
}

function evaluatePollination(ent, world, speciesDef) {
  const pollTargets = speciesDef?.pollination_targets || [];
  if (pollTargets.length === 0) return { type: 'wander' };

  // Find nearest FRUITING flower
  for (const flower of world.entities.values()) {
    if (!flower.isAlive) continue;
    if (!pollTargets.includes(flower.species)) continue;
    if (flower.state !== 'FRUITING') continue;
    const dist = ent.distanceTo(flower);
    if (dist < 15 && dist > 0.5) {
      return {
        type: 'pollinate',
        targetX: flower.x,
        targetZ: flower.z,
        targetId: flower.id,
      };
    }
  }

  // No fruiting flowers — wander to search
  return { type: 'wander' };
}

function evaluateWandering(ent, world) {
  // Reuse existing wander target if still valid and not reached
  if (ent.hasTarget && ent._lastActionType === 'wander') {
    const dx = ent.targetX - ent.x;
    const dz = ent.targetZ - ent.z;
    if (Math.sqrt(dx * dx + dz * dz) > 0.5) {
      // Still have distance to current wander target — keep it
      return { type: 'wander', targetX: ent.targetX, targetZ: ent.targetZ };
    }
  }

  const ml = ent.motionLatent || [0, 0, 0, 0];
  const urgency = Math.abs(ml[0]); // dim 0 = pace/urgency

  // Wander range modulated by urgency — high urgency = tighter wander
  const wanderRange = 2 + (1 - urgency) * 4;
  return {
    type: 'wander',
    targetX: clamp(ent.x + (Math.random() - 0.5) * wanderRange * 2, GRID_SIZE),
    targetZ: clamp(ent.z + (Math.random() - 0.5) * wanderRange * 2, GRID_SIZE),
  };
}

// ─── Action Execution ─────────────────────────────────

/**
 * Execute an action for one entity over one frame.
 * Handles movement toward target and interaction triggers.
 */
function executeAction(ent, action, world, dt, events) {
  if (!action || !action.targetX && action.type !== 'wander') return;

  const ml = ent.motionLatent || [0, 0, 0, 0];
  const urgency = (ml[0] + 1) * 0.5; // normalize to 0..1
  const caution = Math.abs(ml[1]);   // dim 1 = alertness

  // Base speed from species or default
  let baseSpeed = ent.speed || 2.0;

  // Modulate speed by action type and latent urgency
  switch (action.type) {
    case 'flee':
      baseSpeed *= 1.5 + urgency * 0.5; // flee fast
      break;
    case 'hunt':
      baseSpeed *= 1.2 + urgency * 0.3;
      break;
    case 'forage':
      baseSpeed *= 0.8 + urgency * 0.4;
      break;
    case 'drink':
      baseSpeed *= 0.7; // approach water calmly
      break;
    case 'seek_mate':
      baseSpeed *= 0.6 + urgency * 0.3;
      break;
    case 'pollinate':
      baseSpeed *= 0.5 + urgency * 0.3; // butterflies are slow
      break;
    default: // wander
      baseSpeed *= 0.3 + (1 - caution) * 0.4;
  }

  // Move toward target
  const dx = action.targetX - ent.x;
  const dz = action.targetZ - ent.z;
  const dist = Math.sqrt(dx * dx + dz * dz);

  if (dist > 0.1) {
    const step = Math.min(baseSpeed * dt, dist);
    // Add slight curvature based on caution (dim 1) — high caution = more wobble
    const wobble = caution * Math.sin(performance.now() * 0.003 + ent.x) * 0.3;

    ent.x += (dx / dist) * step + wobble * dt;
    ent.z += (dz / dist) * step - wobble * dt;
    ent.x = clamp(ent.x, GRID_SIZE);
    ent.z = clamp(ent.z, GRID_SIZE);

    ent.velocityX = (dx / dist) * baseSpeed;
    ent.velocityZ = (dz / dist) * baseSpeed;
  } else {
    // Arrived at target
    ent.velocityX = 0;
    ent.velocityZ = 0;

    // Check for interaction triggers on arrival (with cooldown)
    checkInteraction(ent, action, world, events);

    // Reset wander target so next frame picks a new one
    if (action.type === 'wander') {
      ent.hasTarget = false;
    }
  }

  ent.targetX = action.targetX;
  ent.targetZ = action.targetZ;
  ent._lastActionType = action.type;
  ent.hasTarget = true;
}

/**
 * Check if an entity should trigger an interaction on target arrival.
 * Reports events upstream for server absorption.
 * Uses per-target cooldown to prevent event spam.
 */
function checkInteraction(ent, action, world, events) {
  const targetEnt = action.targetId ? world.entities.get(action.targetId) : null;
  if (!targetEnt || !targetEnt.isAlive) return;

  const dist = ent.distanceTo(targetEnt);
  if (dist > 3.0) return; // too far for any interaction

  // Cooldown: don't re-interact with same target within 2 seconds
  const now = performance.now();
  const key = `${ent.id}:${action.targetId}`;
  if (!ent._interactionCooldowns) ent._interactionCooldowns = {};
  if (ent._interactionCooldowns[key] && now - ent._interactionCooldowns[key] < 2000) {
    return; // still on cooldown
  }

  switch (action.type) {
    case 'forage':
      if (dist < 2.0 && ent.canConsume) {
        events.push({
          type: 'consumption',
          source_id: ent.id,
          target_id: targetEnt.id,
          position: [targetEnt.x, 0, targetEnt.z],
        });
        ent._interactionCooldowns[key] = now;
      }
      break;

    case 'hunt':
      if (dist < 1.5 && ent.canPredatate) {
        events.push({
          type: 'predation',
          source_id: ent.id,
          target_id: targetEnt.id,
          kill_position: [targetEnt.x, 0, targetEnt.z],
        });
        ent._interactionCooldowns[key] = now;
      }
      break;

    case 'pollinate':
      if (dist < 1.5 && ent.canPollinate) {
        events.push({
          type: 'pollination',
          source_id: ent.id,
          target_id: targetEnt.id,
          position: [targetEnt.x, 0, targetEnt.z],
        });
        ent._interactionCooldowns[key] = now;
      }
      break;

    case 'seek_mate':
      if (dist < 3.0 && ent.reproEligible) {
        events.push({
          type: 'repro',
          parent_id: ent.id,
          offspring_count: 1,
          client_position: [ent.x, 0, ent.z],
        });
        ent._interactionCooldowns[key] = now;
      }
      break;
  }
}

// ─── Helpers ──────────────────────────────────────────

function clamp(v, max) {
  return Math.max(0.5, Math.min(max - 0.5, v));
}

File: client/python/lila_client/agency.py
"""līlā Python Client — Client-side agency engine.

Between server ticks, each mobile entity decides what to do based on:
  - Server intent (state + drives + eligibility flags)
  - Local perception (nearest food, water, threats from world model)
  - Motion latent (modulates speed, hesitation, path curvature)

This is the "body" in "server is nervous system, client is body."
"""

from __future__ import annotations

import math
import random
import time
from typing import Any

from .constants import GRID_SIZE
from .world_model import WorldEntity


def step_agency(world, dt: float) -> list[dict]:
    """Run one frame of local agency for all mobile entities.

    Called every render frame (~60 Hz). Returns client-reported events
    to send upstream via heartbeat.
    """
    events: list[dict] = []

    for ent in world.entities.values():
        if not ent.is_mobile_consumer or not ent.is_alive:
            continue

        # Update speed from species definition (once)
        if not getattr(ent, "_speed_set", False):
            defn = world.species_defs.get(ent.species, {})
            ent.speed = defn.get("movement_speed", 2.0)
            ent._speed_set = True

        # Evaluate behavior
        action = evaluate_behavior(ent, world)

        # Execute the action
        execute_action(ent, action, world, dt, events)

    return events


# ─── Behavior Evaluators ──────────────────────────────────────────────────


def evaluate_behavior(ent: WorldEntity, world) -> dict[str, Any]:
    """Evaluate what an entity should do based on its intent and local perception."""
    state = ent.state
    drive = ent.drive or {}
    species_def = world.species_defs.get(ent.species, {})

    # ── Fleeing (highest priority) ──
    if state == "FLEEING":
        return evaluate_fleeing(ent, world, species_def)

    # ── Drinking ──
    if state == "DRINKING" or (ent.can_drink and drive.get("hydration", 1.0) < 0.3):
        return evaluate_drinking(ent, world)

    # ── Reproduction seeking ──
    if ent.repro_eligible and drive.get("reproductive_drive", 0) > 0.5:
        mate_action = evaluate_mate_seeking(ent, world)
        if mate_action:
            return mate_action

    # ── Foraging / Herbivory ──
    if state == "FORAGING" and ent.can_consume:
        return evaluate_foraging(ent, world, species_def)

    # ── Hunting / Predation ──
    if (state in ("HUNTING", "FORAGING")) and ent.can_predate:
        return evaluate_hunting(ent, world, species_def)

    # ── Pollination ──
    if ent.can_pollinate and species_def.get("is_pollinator"):
        return evaluate_pollination(ent, world, species_def)

    # ── Resting / Idle — wander with latent-modulated style ──
    return evaluate_wandering(ent, world)


def evaluate_fleeing(ent: WorldEntity, world, species_def: dict) -> dict:
    """Flee from nearest threat."""
    flee_targets = species_def.get("flee_targets", [])
    if not flee_targets:
        return evaluate_wandering(ent, world)

    nearest_threat: WorldEntity | None = None
    best_dist_sq = float("inf")
    for other in world.entities.values():
        if not other.is_alive:
            continue
        other_def = world.species_defs.get(other.species, {})
        if not other_def or other.species not in flee_targets:
            continue
        d2 = ent.dist_sq_to(other)
        if d2 < best_dist_sq:
            best_dist_sq = d2
            nearest_threat = other

    if nearest_threat and best_dist_sq < 400:  # ~20 world units sensory range²
        dx = ent.x - nearest_threat.x
        dz = ent.z - nearest_threat.z
        dist = math.sqrt(dx * dx + dz * dz) or 1
        return {
            "type": "flee",
            "target_x": clamp(ent.x + (dx / dist) * 8, GRID_SIZE),
            "target_z": clamp(ent.z + (dz / dist) * 8, GRID_SIZE),
        }

    return evaluate_wandering(ent, world)


def evaluate_drinking(ent: WorldEntity, world) -> dict:
    """Approach nearest water source."""
    water = world.find_nearest_water(ent.x, ent.z)
    if water:
        wx, _, wz = water["position"]
        r = water.get("radius", 1)
        dx = wx - ent.x
        dz = wz - ent.z
        dist = math.sqrt(dx * dx + dz * dz) or 1
        approach_r = max(r - 0.5, 0.3)
        return {
            "type": "drink",
            "target_x": clamp(wx - (dx / dist) * approach_r, GRID_SIZE),
            "target_z": clamp(wz - (dz / dist) * approach_r, GRID_SIZE),
        }
    return evaluate_wandering(ent, world)


def evaluate_mate_seeking(ent: WorldEntity, world) -> dict | None:
    """Seek a mate of the same species."""
    best: WorldEntity | None = None
    best_dist = 15.0  # sensory range
    for other in world.entities.values():
        if other.id == ent.id:
            continue
        if other.species != ent.species:
            continue
        if not other.is_alive:
            continue
        d = ent.distance_to(other)
        if d < best_dist:
            best_dist = d
            best = other

    if best:
        return {
            "type": "seek_mate",
            "target_x": best.x,
            "target_z": best.z,
            "target_id": best.id,
        }
    return None


def evaluate_foraging(ent: WorldEntity, world, species_def: dict) -> dict:
    """Find nearest food item to approach."""
    diet_order = species_def.get("diet_order", [])

    for food_species in diet_order:
        if isinstance(food_species, (list, tuple)):
            food_species = food_species[0]
        food = world.find_nearest_species(ent.x, ent.z, [food_species], ent.id)
        if food and ent.distance_to(food) < 15:
            return {
                "type": "forage",
                "target_x": food.x,
                "target_z": food.z,
                "target_id": food.id,
            }

    return evaluate_wandering(ent, world)


def evaluate_hunting(ent: WorldEntity, world, species_def: dict) -> dict:
    """Find nearest prey to hunt."""
    diet_order = species_def.get("diet_order", [])
    prey_species = [s[0] if isinstance(s, (list, tuple)) else s for s in diet_order]

    if prey_species:
        prey = world.find_nearest_species(ent.x, ent.z, prey_species, ent.id)
        if prey and ent.distance_to(prey) < 15:
            return {
                "type": "hunt",
                "target_x": prey.x,
                "target_z": prey.z,
                "target_id": prey.id,
            }

    return evaluate_wandering(ent, world)


def evaluate_pollination(ent: WorldEntity, world, species_def: dict) -> dict:
    """Find nearest fruiting flower to pollinate."""
    poll_targets = species_def.get("pollination_targets", [])
    if not poll_targets:
        return evaluate_wandering(ent, world)

    for flower in world.entities.values():
        if not flower.is_alive:
            continue
        if flower.species not in poll_targets:
            continue
        if flower.state != "FRUITING":
            continue
        dist = ent.distance_to(flower)
        if 0.5 < dist < 15:
            return {
                "type": "pollinate",
                "target_x": flower.x,
                "target_z": flower.z,
                "target_id": flower.id,
            }

    return evaluate_wandering(ent, world)


def evaluate_wandering(ent: WorldEntity, world) -> dict:
    """Pick a wander target near the entity, modulated by motion latent."""
    # Reuse existing wander target if still valid and not reached
    if getattr(ent, "has_target", False) and getattr(ent, "_last_action_type", "") == "wander":
        dx = ent.target_x - ent.x
        dz = ent.target_z - ent.z
        if math.sqrt(dx * dx + dz * dz) > 0.5:
            return {"type": "wander", "target_x": ent.target_x, "target_z": ent.target_z}

    ml = ent.motion_latent or [0, 0, 0, 0]
    urgency = abs(ml[0])  # dim 0 = pace/urgency
    wander_range = 2 + (1 - urgency) * 4

    return {
        "type": "wander",
        "target_x": clamp(ent.x + (random.random() - 0.5) * wander_range * 2, GRID_SIZE),
        "target_z": clamp(ent.z + (random.random() - 0.5) * wander_range * 2, GRID_SIZE),
    }


# ─── Action Execution ─────────────────────────────────────────────────────


def execute_action(
    ent: WorldEntity,
    action: dict[str, Any],
    world,
    dt: float,
    events: list[dict],
) -> None:
    """Execute an action for one entity over one frame."""
    if not action or ("target_x" not in action and action.get("type") != "wander"):
        return

    ml = ent.motion_latent or [0, 0, 0, 0]
    urgency = (ml[0] + 1) * 0.5  # normalize to 0..1
    caution = abs(ml[1])  # dim 1 = alertness

    base_speed = ent.speed or 2.0

    # Modulate speed by action type
    speed_mod = {
        "flee": 1.5 + urgency * 0.5,
        "hunt": 1.2 + urgency * 0.3,
        "forage": 0.8 + urgency * 0.4,
        "drink": 0.7,
        "seek_mate": 0.6 + urgency * 0.3,
        "pollinate": 0.5 + urgency * 0.3,
    }.get(action.get("type", "wander"), 0.3 + (1 - caution) * 0.4)

    base_speed *= speed_mod

    target_x = action.get("target_x", ent.x)
    target_z = action.get("target_z", ent.z)

    # Move toward target
    dx = target_x - ent.x
    dz = target_z - ent.z
    dist = math.sqrt(dx * dx + dz * dz)

    if dist > 0.1:
        step = min(base_speed * dt, dist)
        # Add slight curvature based on caution — high caution = more wobble
        wobble = caution * math.sin(time.monotonic() * 3 + ent.x) * 0.3

        ent.x += (dx / dist) * step + wobble * dt
        ent.z += (dz / dist) * step - wobble * dt
        ent.x = clamp(ent.x, GRID_SIZE)
        ent.z = clamp(ent.z, GRID_SIZE)

        ent.velocity_x = (dx / dist) * base_speed
        ent.velocity_z = (dz / dist) * base_speed

        # Lerp facing angle toward travel direction (smooth rotation)
        target_angle = math.atan2(dz, dx)
        ent.facing_angle = _lerp_angle(ent.facing_angle, target_angle, 0.15)
    else:
        # Arrived at target
        ent.velocity_x = 0
        ent.velocity_z = 0

        # Check for interaction triggers on arrival
        _check_interaction(ent, action, world, events)

        # Reset wander target so next frame picks a new one
        if action.get("type") == "wander":
            ent.has_target = False

    ent.target_x = target_x
    ent.target_z = target_z
    ent._last_action_type = action.get("type", "wander")
    ent.has_target = True


def _check_interaction(
    ent: WorldEntity,
    action: dict[str, Any],
    world,
    events: list[dict],
) -> None:
    """Check if an entity should trigger an interaction on target arrival."""
    target_id = action.get("target_id")
    if not target_id:
        return

    target_ent = world.entities.get(target_id)
    if not target_ent or not target_ent.is_alive:
        return

    dist = ent.distance_to(target_ent)
    if dist > 3.0:
        return

    # Cooldown: don't re-interact with same target within 2 seconds
    now = time.monotonic()
    cooldowns = getattr(ent, "_interaction_cooldowns", {})
    key = f"{ent.id}:{target_id}"
    if cooldowns.get(key, 0) > now - 2.0:
        return

    action_type = action.get("type", "")

    if action_type == "forage" and dist < 2.0 and ent.can_consume:
        events.append({
            "type": "consumption",
            "source_id": ent.id,
            "target_id": target_id,
            "position": [target_ent.x, 0, target_ent.z],
        })
        cooldowns[key] = now

    elif action_type == "hunt" and dist < 1.5 and ent.can_predate:
        events.append({
            "type": "predation",
            "source_id": ent.id,
            "target_id": target_id,
            "kill_position": [target_ent.x, 0, target_ent.z],
        })
        cooldowns[key] = now

    elif action_type == "pollinate" and dist < 1.5 and ent.can_pollinate:
        events.append({
            "type": "pollination",
            "source_id": ent.id,
            "target_id": target_id,
            "position": [target_ent.x, 0, target_ent.z],
        })
        cooldowns[key] = now

    elif action_type == "seek_mate" and dist < 3.0 and ent.repro_eligible:
        events.append({
            "type": "repro",
            "parent_id": ent.id,
            "offspring_count": 1,
            "client_position": [ent.x, 0, ent.z],
        })
        cooldowns[key] = now

    ent._interaction_cooldowns = cooldowns


# ─── Helpers ────────────────────────────────────────────────────────────────


def clamp(v: float, max_val: int) -> float:
    return max(0.5, min(max_val - 0.5, v))


def _lerp_angle(a: float, b: float, t: float) -> float:
    """Spherical lerp between two angles (handles wrapping at ±π)."""
    # Normalize difference to [-π, π]
    diff = math.atan2(math.sin(b - a), math.cos(b - a))
    return a + diff * t
